### PR TITLE
Add a basic type stub file

### PR DESCRIPTION
**Pull request by chrisnorman7**
Describe the feature and/or bug fix:

I've added `synthizer.pyi` to the `bindings/python` directory. I'm unsure how to make it be shipped with Synthizer proper, but at least now you've got something to look at.

This is for issue #9

### CODE REVIEW

File: bindings/python/synthizer.pyi
Comment (camlorn) [inline bindings/python/synthizer.pyi:14]: You do this on all the enums, where you lift the constant out of the enum as well as putting it under the enum.  Given that Synthizer is `synthizer.LogLevel.STDERR` etc. aren't these top-level definitions incorrect/redundant?

File: bindings/python/synthizer.pyi
Comment (camlorn) [inline bindings/python/synthizer.pyi:65]: This is a context manager, not an iterator.

File: bindings/python/synthizer.pyi
Comment (camlorn) [inline bindings/python/synthizer.pyi:83]: This isn't part of the public API, and shouldn't be used by users. Is it necessary to type it?

File: bindings/python/synthizer.pyi
Comment (camlorn) [inline bindings/python/synthizer.pyi:119]: Please don't include the defaults on properties unless this is necessary for mypy for some reason.  I doubt they will be kept up to date long term.  I don't want to have to audit this before every release that might change one.

**Review by camlorn — CHANGES_REQUESTED** [review]
I'll need you to work out how to get this shipping before I merge.  Also:

- What's with the double blank lines between classes and enums?  is that typical Mypy formatting?
- Is there an autoformatter?
- Do we need to include defaults here, or is it sufficient just to say there is one?  I noticed some places you used `...`, and (as pointed out below) I don't want to have to be auditing this to that level of detail if it can be avoided.

File: bindings/python/synthizer.pyi
Comment (chrisnorman7) [reply bindings/python/synthizer.pyi:14]: Given that:

```python
synthizer.ERROR is synthizer.LogLevel.ERROR
```

is `True`, I thought they were OK.

File: bindings/python/synthizer.pyi
Comment (chrisnorman7) [reply bindings/python/synthizer.pyi:65]: I was working according to [https://stackoverflow.com/questions/49733699/python-type-hints-and-context-managers](this question), where it mentions typing the function as an iterator.

That said, there is also `typing.ContextManager`, which may be more appropriate. I'll make that change.

File: bindings/python/synthizer.pyi
Comment (chrisnorman7) [reply bindings/python/synthizer.pyi:119]: That's fine. I'll go through and remove them.

**Comment by chrisnorman7** [comment]
So it seems black works fine on pyi files (I just ran it).

Only reason I included defaults was because VS Code shows them as hints upon occasion, and I'm not sure that would be the case if they weren't there.

Also, I have no clue how to ship the file. I'll look more into that once I've made the code changes you requested.

**Comment by chrisnorman7** [comment]
OK, this all works now on my system. It seems to install correctly and everything.

I don't particularly like not having defaults for function arguments, would you be OK with them being added back in when the API stabilises?

**Comment by ahicks92** [comment]
I'm going to punt on defaults on function arguments for now.  I would have assumed that VSCode should be picking them up anyway, but perhaps Cython isn't exposing the metadata.  I could have sworn I saw it working with Numpy but I wasn't looking closely.

I don't object in principle, but I do object in the sense that a wrong default there is worse than not knowing what the default is.

Let's see how things fall out at 1.0.

I still see a bunch of top-level renames of the enums.  before proceeding with this, I'd like to understand why you think that's correct?  You can't do `synthizer.STDERR` for example.  It has to be `synthizer.LogLevel.STDERR`.  Doesn't the top-level definition imply that the module has an import of the same name?

In general this looks good though.  Thanks for putting in the effort.  Just want to understand the enum thing.

**Comment by chrisnorman7** [comment]
Actually, there are top-level definitions:

```python
Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64
bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import synthizer as sz
>>> sz.STDERR
<LoggingBackend.STDERR: 0>
```

Cheers,

Take care,

Chris Norman

**Comment by ahicks92** [comment]
There shouldn't be top-level definitions.  They aren't in the source code.  Either Cython is doing something I don't want it doing, or I'm using it wrong.  I will figure out which.  Don't rely on those definitions.

I'll merge this in after I finish migrating CI to GitHub actions.  Not going to keep it out over points as minor as these.  Thanks for the work.

**Pull request merged**
